frame/support/procedural/src/construct_runtime/mod.rs

```diff
@@ -425,3 +450,67 @@ fn decl_integrity_test(scrate: &TokenStream2) -> TokenStream2 {
 		}
 	)
 }
+
+fn get_available_indices(modules: &Vec<ModuleDeclaration>) -> Vec<u8> {
+	let reserved_indices = modules.iter()
+		.filter_map(|module| module.index)
+		.collect::<Vec<_>>();
+
+	(0..u8::max_value())
+	.filter(|i| !reserved_indices.contains(i))
+	.collect::<Vec<_>>()
+}
+
+/// Check:
+/// * check system module is defined
+/// * there is less than 256 modules
+/// * no module use index 0 or except system
+/// * module indices don't conflict.
+///
+/// returns system module ident
+fn check_modules(
+	modules: &Vec<ModuleDeclaration>,
+	modules_span: proc_macro2::Span,
+) -> Result<syn::Ident> {
+	// Assert we have system module declared
+	let system_module = modules.iter()
+		.find(|decl| decl.name == SYSTEM_MODULE_NAME)
+		.ok_or_else(|| syn::Error::new(
+			modules_span,
+			"`System` module declaration is missing. \
+			 Please add this line: `System: frame_system::{Module, Call, Storage, Config, Event<T>},`",
+		))?;
+
+	// Assert no more than 256 modules
+	if modules.len() > u8::max_value() as usize {
+		let msg = "Too many modules defined, only 256 modules is currently allowed";
+		return Err(syn::Error::new(modules_span, msg));
+	}
+
+	// No module use index 0 explicity (or it is system)
+	if let Some(module) = modules.iter()
+		.find(|module| {
+			module.index.map_or(false, |i| i == 0) && module.name != SYSTEM_MODULE_NAME
+		})
+	{
+		let msg = format!(
+			"Only system module is allowed to be defined at index `0`, this is because of origin \
```

Comment:
Why is this because of origin encoding?

Reply:
First note: I tried not to break anything, if we want to break current encoding of origin, then other solution arise.

currently origin encodes system variant at index 0 (no matter where is system defined in construct_runtime).
I though to avoid more confusion in how origin enum is generated I would keep this behavior.
but indeed we can do: if another pallet is using index 0 then system is assigned the first available implicit index.

But if we do as you propose `Set the index for every module and don't rely anymore on the implicit index of the module`. Then we can do a breaking change and fix origin encoding. I.e. make system variant being at the index defined.

Reply:
Where is the encoding of the `Origin` important?

Reply:
because it is stored in scheduler

Reply:
Ahh okay. I don't have any strong opinion on breaking origin or not.

Reply:
we are ok break origin, so this is removed